adapters/cisco_isr/do_cmd_enroll_certificate.php

```diff
@@ -0,0 +1,223 @@
+$rsakeypairlabel='RSA4096';
+
+// We check that we have parameters
+if (!isset($optional_params))
+{
+    $ret = ERR_VERB_BAD_PARAM;
+    sms_send_user_error($sms_csp, $sdid, '["Missing params"]', $ret);
+    return $ret;
+}
+$ret = SMS_OK;
+
+// We decode parameters
+$params = json_decode($optional_params, true);
+debug_dump($params, 'PARAMS');
+
+if (is_null($params))
+{
+    switch (json_last_error()) {
+        case JSON_ERROR_NONE:
+            $error_msg = 'No error';
+            break;
+        case JSON_ERROR_DEPTH:
+            $error_msg = 'Maximum depth reached';
+            break;
+        case JSON_ERROR_STATE_MISMATCH:
+            $error_msg = 'Mismatch of modes or underflow';
+            break;
+        case JSON_ERROR_CTRL_CHAR:
+            $error_msg = 'Contains bad control chars';
+            break;
+        case JSON_ERROR_SYNTAX:
+            $error_msg = 'JSON syntax error';
+            break;
+        case JSON_ERROR_UTF8:
+            $error_msg = 'Contains wrong UTF-8 chars, probably an encoding issue';
+            break;
+        default:
+            $error_msg = 'Unknown error';
+        break;
+    }
+    $ret = ERR_VERB_BAD_PARAM;
+    sms_log_error("$error_msg : $optional_params\n");
+    sms_send_user_error($sms_csp, $sdid, "[\"$error_msg\":$optional_params]", $ret);
+    return $ret;
+}
+
+// We check that we have all the requested parameters
+foreach ($req_params as $req_param)
+{
+    if (empty($params[$req_param]))
+    {
+        $ret = ERR_VERB_BAD_PARAM;
+        $error_msg = 'Missing parameter';
+        sms_log_error("$error_msg : $req_param in $optional_params\n");
+        sms_send_user_error($sms_csp, $sdid, "[\"$error_msg\":\"$req_param\", \"parameters\":$optional_params]", $ret);
+        return $ret;
+    }
+}
+
+// ====== Real work ============================================================
+global $sms_sd_ctx;
+$network = get_network_profile();
+$SD = &$network->SD;
+
+// Lock
+$ret = sms_sd_lock($sms_csp, $sms_sd_info);
+if ($ret !== 0)
+{
+  sms_send_user_error($sms_csp, $sdid, "", $ret);
+  sms_close_user_socket($sms_csp);
+  return SMS_OK;
+}
+
+$ret = cisco_isr_connect();
+if ($ret !== SMS_OK)
+{
+  sms_log_error(__FILE__.':'.__LINE__.": cisco_isr_connect() failed\n");
+  sms_send_user_error($sms_csp, $sdid, "", $ret);
+  return $ret;
+}
+
+$on_error_fct = 'cisco_isr_disconnect';
+$result = '';
+
+sendexpectnobuffer(__FILE__ . ':' . __LINE__, $sms_sd_ctx, 'conf t', '(config)#');
+
+// we remove possibly existing one, otherwise it will generate an error
+unset($tab);
+$tab[0] = "Can't find policy PKI";
+$tab[1] = 'sure you want to do this? [yes/no]:';
+$index = sendexpect(__FILE__.':'.__LINE__, $sms_sd_ctx, 'no crypto pki trustpoint PKI', $tab);
+if ($index == 1)
+{
+  sendexpectone(__FILE__.':'.__LINE__, $sms_sd_ctx, 'yes', '(config)#');
+}
+
+sendexpectnobuffer(__FILE__ . ':' . __LINE__, $sms_sd_ctx, '!configure trustpoint/PKI end point', '(config)#');
```

Comment:
Sending comment has no effect, so this line 130 is useless. It could be replaced by a comment in the code.